File: 01_ctrlGPIO/06_hcsr04/hc-sr04.c
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <wiringPi.h>

#define TRIG 24	
#define ECHO 25

int trig(void)
{
	digitalWrite( TRIG, HIGH );
	delayMicroseconds( 10 );
	digitalWrite( TRIG, LOW  );
}

unsigned long echo(void)
{
	unsigned long time = 0;
	
	while( digitalRead( ECHO ) == LOW  );
	time = micros();
	while( digitalRead( ECHO ) == HIGH );
	time = micros() - time;
	
	if( time >= 240 && time <= 23000 )	return time;
	
	else	return 0;
}

int main (void)
{
	int data;

	unsigned long echo_time = 0;
	unsigned long dist = 0;
  
	wiringPiSetup();
	
	pinMode( ECHO, INPUT  );
	pinMode( TRIG, OUTPUT );
	digitalWrite( TRIG, LOW );
  
	printf ("\nHC-SR04 Test~\n");
  
	while(1)
	{
		trig();
		echo_time = echo();
		
		if( echo_time != 0 )
		{
			// 340000mm / 1000000us = 2 * dist / echo_time
			// 340000 * echo_time = 2 * dist * 1000000
			// dist = 340000 * echo_time / 2000000
			// dist = 17 * echo_time / 100
			dist = 17 * echo_time / 100;
			printf("distance = %4d (mm)\n",dist);
		}
		else
			printf("out of range!!!\n",dist);
		
		delay(1000);
	}
	return 0;
}

File: 01_ctrlGPIO/06_hcsr04/hc-sr04.py
import RPi.GPIO as GPIO
import time

GPIO.setmode(GPIO.BCM)

TRIG = 19
ECHO = 26

print "Measurement Distance"

GPIO.setup(TRIG,GPIO.OUT)
GPIO.setup(ECHO,GPIO.IN)

GPIO.output(TRIG, False)
time.sleep(1)

try:
  while True:
    GPIO.output(TRIG, True)
    time.sleep(0.000010)          # 10 micro sec
    GPIO.output(TRIG, False)

    while GPIO.input(ECHO)==0:
      echo_start = time.time()
      
    while GPIO.input(ECHO)==1:
      echo_end = time.time()
      
    echo_time = echo_end - echo_start
    dist = echo_time * 170000
    dist = round(dist, 2)
    print "Distance = ",dist,"mm"

    time.sleep(1);

except KeyboardInterrupt:
  GPIO.cleanup()